# http://redmine.encodedcc.org/issues/4987

label_mapping = {
    '/analysis-steps/alignment-pooling-step/': 'alignment-pooling-step',
	'/analysis-steps/alignment-subsampling-step/': 'alignment-subsampling-step',
	'/analysis-steps/atac-seq-alignment-step-v-1/': 'atac-seq-alignment-step',
	'/analysis-steps/atac-seq-peak-calling-step-v_1/': 'atac-seq-peak-calling-step',
	'/analysis-steps/atac-seq-peak-filtering-step-v_1/': 'atac-seq-peak-filtering-step',
	'/analysis-steps/atac-seq-region-calling-step-v-1/': 'atac-seq-region-calling-step',
	'/analysis-steps/atac-seq-region-filter-step-v-1/': 'atac-seq-region-filter-step',
	'/analysis-steps/atac-seq-signal-gen-step-v-1/': 'atac-seq-signal-gen-step',
	'/analysis-steps/bed-2-bigbed/': 'bed-2-bigbed-step',
	'/analysis-steps/broad_frip_seq_cuffdiff_step/': 'broad-frip-seq-cuffdiff-step',
	'/analysis-steps/broad_frip_seq_fastq_alignment_step/': 'broad-frip-seq-fastq-alignment-step',
	'/analysis-steps/bwa-alignment-step-v-1/': 'bwa-alignment-step',
	'/analysis-steps/bwa-indexing-step-v-1/': 'bwa-indexing-step',
	'/analysis-steps/bwa-raw-alignment-step-v-1/': 'bwa-raw-alignment-step',
	'/analysis-steps/celltype-cre-identification/': 'celltype-cre-identification-step',
	'/analysis-steps/chia_pet_chromatin_interactions_calling_step/': 'chia-pet-chromatin-interactions-calling-step',
	'/analysis-steps/chia_pet_haplotype_specific_step/': 'chia-pet-haplotype-specific-step',
	'/analysis-steps/chia_pet_loop_calling_step/': 'chia-pet-loop-calling-step',
	'/analysis-steps/chia_pet_peak_calling_step/': 'chia-pet-peak-calling-step',
	'/analysis-steps/chia_pet_preprocessing_step/': 'chia-pet-preprocessing-step',
	'/analysis-steps/cluster-filter/': 'cluster-filter-step',
	'/analysis-steps/control-alignment-subsampling-step/': 'control-alignment-subsampling-step',
	'/analysis-steps/dac-enhancer-like-ranking-method/': 'dac-enhancer-like-ranking-method-step',
	'/analysis-steps/dac-promotor-like-ranking-method/': 'dac-promotor-like-ranking-method-step',
	'/analysis-steps/dme-align-se-v-2/': 'deleted-dme-align-se-step',
	'/analysis-steps/dme-index-bismark-v-2_deleted/': 'deleted-dme-index-bismark-step',
	'/analysis-steps/index-star-v-2-0/': 'deleted-lrna-index-star-step',
	'/analysis-steps/lrna-pe-star-alignment-step-v-2-0/': 'deleted-lrna-pe-star-alignment-step',
	'/analysis-steps/lrna-pe-star-stranded-signal-step-v-2-0/': 'deleted-lrna-pe-star-stranded-signal-step',
	'/analysis-steps/lrna-pe-star-stranded-signals-for-tophat-step-v-2-0/': 'deleted-lrna-pe-star-stranded-signals-for-tophat-step',
	'/analysis-steps/lrna-se-star-alignment-step-v-2-0/': 'deleted-lrna-se-star-alignment-step',
	'/analysis-steps/lrna-se-star-unstranded-signal-step-v-2-0/': 'deleted-lrna-se-star-unstranded-signal-step',
	'/analysis-steps/lrna-se-star-unstranded-signals-for-tophat-step-v-2-0/': 'deleted-lrna-se-star-unstranded-signals-for-tophat-step',
	'/analysis-steps/rampage-grit-peak-calling-step-v-1-1/': 'deleted-rampage-grit-peak-calling-step',
	'/analysis-steps/dme-align-pe-v-1/': 'dme-align-pe-step',
	'/analysis-steps/mott-trim-align-bismark-v-1-0/': 'dme-align-se-step',
	'/analysis-steps/dme-extract-pe-v-1/': 'dme-extract-pe-step',
	'/analysis-steps/dme-extract-se-v-2/': 'dme-extract-se-step',
	'/analysis-steps/wgbs-bismark-indexing-step-v-1/': 'dme-index-bismark-bowtie1-step',
	'/analysis-steps/dme-index-bismark-bowtie2-v-1/': 'dme-index-bismark-bowtie2-step',
	'/analysis-steps/dme-rep-corr-v-1/': 'dme-rep-corr-pe-step',
	'/analysis-steps/dme-rep-corr-alt-v-1/': 'dme-rep-corr-se-step',
	'/analysis-steps/dnase-align-bwa-pe-v-1/': 'dnase-align-bwa-pe-step',
	'/analysis-steps/dnase-align-bwa-se-v-1/': 'dnase-align-bwa-se-step',
	'/analysis-steps/dnase-bed-to-bigbed/': 'dnase-bed-to-bigbed-step',
	'/analysis-steps/dnase-call-hotspots-v-1/': 'dnase-call-hotspots-pe-step',
	'/analysis-steps/dnase-call-hotspots-alt-v-1/': 'dnase-call-hotspots-se-step',
	'/analysis-steps/dnase-eval-bam-v-1/': 'dnase-eval-bam-pe-step',
	'/analysis-steps/dnase-eval-bam-alt-v-1/': 'dnase-eval-bam-se-step',
	'/analysis-steps/dnase-filter-pe-v-1/': 'dnase-filter-pe-step',
	'/analysis-steps/dnase-filter-se-v-1/': 'dnase-filter-se-step',
	'/analysis-steps/dnase-index-bwa-v-1/': 'dnase-index-bwa-step',
	'/analysis-steps/dnase-rep-corr-v-1/': 'dnase-rep-corr-pe-step',
	'/analysis-steps/dnase-rep-corr-alt-v-1/': 'dnase-rep-corr-se-step',
	'/analysis-steps/dnase-seq-mapping-peak-calling-step-v-1/': 'dnase-seq-mapping-peak-calling-step',
	'/analysis-steps/dnase-seq-mapping-step-v2/': 'dnase-seq-mapping-step',
	'/analysis-steps/dnase-seq-peak-calling-step-v2/': 'dnase-seq-peak-calling-step',
	'/analysis-steps/eclip_bigbed_to_bed_step/': 'eclip-bigbed-to-bed-step',
	'/analysis-steps/eclip_makebigwig_step/': 'eclip-makebigwig-step',
	'/analysis-steps/eclip_normalization_step/': 'eclip-normalization-step',
	'/analysis-steps/eclip_trimming_mapping_step/': 'eclip-trimming-mapping-step',
	'/analysis-steps/encode2-dgf-analysis/': 'encode2-dgf-analysis-step',
	'/analysis-steps/encode2-dnase-analysis/': 'encode2-dnase-analysis-step',
	'/analysis-steps/generalized-cre-identification/': 'generalized-cre-identification-step',
	'/analysis-steps/ggr_cl1_atac_seq_bam_to_bed_step/': 'ggr-cl1-atac-seq-bam-to-bed-step',
	'/analysis-steps/ggr_cl1_atac_seq_idr_step/': 'ggr-cl1-atac-seq-idr-step',
	'/analysis-steps/ggr_cl1_atac_seq_mapping_step/': 'ggr-cl1-atac-seq-mapping-step',
	'/analysis-steps/ggr_cl1_chip_seq_mapping_step/': 'ggr-cl1-chip-seq-mapping-step',
	'/analysis-steps/ggr_cl1_chip_seq_quantification_step/': 'ggr-cl1-chip-seq-quantification-step',
	'/analysis-steps/ggr_cl1_rna_seq_quantification_step/': 'ggr-cl1-rna-seq-quantification-step',
	'/analysis-steps/ggr_cl1_rna_seq_trimming_mapping_filtering_step/': 'ggr-cl1-rna-seq-trimming-mapping-filtering-step',
	'/analysis-steps/ggr_lg1_atac_seq_bet_to_bigwig_step/': 'ggr-lg1-atac-seq-bet-to-bigwig-step',
	'/analysis-steps/ggr_lg1_atac_seq_fastq_align_n_filter_step/': 'ggr-lg1-atac-seq-fastq-align-n-filter-step',
	'/analysis-steps/ggr_lg1_atac_seq_peak_calling_step/': 'ggr-lg1-atac-seq-peak-calling-step',
	'/analysis-steps/ggr_lg1_chip_seq_bam_to_bigwig_step/': 'ggr-lg1-chip-seq-bam-to-bigwig-step',
	'/analysis-steps/ggr_lg1_chip_seq_fastq_alignment_step/': 'ggr-lg1-chip-seq-fastq-alignment-step',
	'/analysis-steps/ggr_lg1_chip_seq_peak_calling_step/': 'ggr-lg1-chip-seq-peak-calling-step',
	'/analysis-steps/ggr_lg1_rna_seq_bam2bigwig_step/': 'ggr-lg1-rna-seq-bam2bigwig-step',
	'/analysis-steps/ggr_lg1_rna_seq_fastq_filtration_step/': 'ggr-lg1-rna-seq-fastq-filtration-step',
	'/analysis-steps/ggr_lg1_rna_seq_mapping_step/': 'ggr-lg1-rna-seq-mapping-step',
	'/analysis-steps/ggr_lg1_rna_seq_rsem_step/': 'ggr-lg1-rna-seq-rsem-step',
	'/analysis-steps/ggr_lg2_rna_seq_bam2bigwig_step/': 'ggr-lg2-rna-seq-bam2bigwig-step',
	'/analysis-steps/ggr_lg2_rna_seq_fastq_filtration_bowtie_step/': 'ggr-lg2-rna-seq-fastq-filtration-bowtie-step',
	'/analysis-steps/ggr_lg2_rna_seq_fastq_filtration_tophat_step/': 'ggr-lg2-rna-seq-fastq-filtration-tophat-step',
	'/analysis-steps/ggr_lg2_rna_seq_rsem_step/': 'ggr-lg2-rna-seq-rsem-step',
	'/analysis-steps/ggr_tr1_chip_seq_mapping_filtering_step/': 'ggr-tr1-chip-seq-mapping-filtering-step',
	'/analysis-steps/ggr_tr1_chip_seq_peak_calling_step/': 'ggr-tr1-chip-seq-peak-calling-step',
	'/analysis-steps/ggr_tr1_chip_seq_quantification_step/': 'ggr-tr1-chip-seq-quantification-step',
	'/analysis-steps/ggr_tr1_dnase_seq_mapping_filtering_step/': 'ggr-tr1-dnase-seq-mapping-filtering-step',
	'/analysis-steps/ggr_tr1_dnase_seq_peak_calling_step/': 'ggr-tr1-dnase-seq-peak-calling-step',
	'/analysis-steps/ggr_tr1_dnase_seq_quantification_step/': 'ggr-tr1-dnase-seq-quantification-step',
	'/analysis-steps/ggr_tr1_hic_loop_calling_step/': 'ggr-tr1-hic-loop-calling-step',
	'/analysis-steps/ggr_tr1_hic_mapping_step/': 'ggr-tr1-hic-mapping-step',
	'/analysis-steps/ggr_tr1_hic_tad_calling_step/': 'ggr-tr1-hic-tad-calling-step',
	'/analysis-steps/ggr_tr1_rna_seq_gene_quantification_step/': 'ggr-tr1-rna-seq-gene-quantification-step',
	'/analysis-steps/ggr_tr1_rna_seq_signal_generation_step/': 'ggr-tr1-rna-seq-signal-generation-step',
	'/analysis-steps/ggr_tr1_rna_seq_trimming_mapping_step/': 'ggr-tr1-rna-seq-trimming-mapping-step',
	'/analysis-steps/hic-liftover-step/': 'hic-liftover-step',
	'/analysis-steps/histone-overlap-peaks-step-v-1/': 'histone-overlap-peaks-step',
	'/analysis-steps/histone-peak-calling-step-v-1/': 'histone-peak-calling-step',
	'/analysis-steps/histone-peaks-to-bigbed-step-v-1/': 'histone-peaks-to-bigbed-step',
	'/analysis-steps/histone-replicated-peaks-to-bigbed-step-v-1/': 'histone-replicated-peaks-to-bigbed-step',
	'/analysis-steps/histone-unreplicated-partition-concordance-peaks-to-bigbed-step/': 'histone-unreplicated-partition-concordance-peaks-to-bigbed-step',
	'/analysis-steps/histone-unreplicated-partition-concordance-step/': 'histone-unreplicated-partition-concordance-step',
	'/analysis-steps/histone-unreplicated-peak-calling-step/': 'histone-unreplicated-peak-calling-step',
	'/analysis-steps/histone-unreplicated-peaks-to-bigbed-step/': 'histone-unreplicated-peaks-to-bigbed-step',
	'/analysis-steps/kundaje-lab-atac-seq-filtered-peaks-conversion-single-rep-v1/': 'kundaje-lab-atac-seq-filtered-peaks-conversion-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-filtered-peaks-conversion-v1/': 'kundaje-lab-atac-seq-filtered-peaks-conversion-step',
	'/analysis-steps/kundaje-lab-atac-seq-idr-peaks-conversion-v1/': 'kundaje-lab-atac-seq-idr-peaks-conversion-step',
	'/analysis-steps/kundaje-lab-atac-seq-idr-v1/': 'kundaje-lab-atac-seq-idr-step',
	'/analysis-steps/kundaje-lab-atac-seq-overlap-peaks-single-rep-v1/': 'kundaje-lab-atac-seq-overlap-peaks-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-overlap-peaks-v1/': 'kundaje-lab-atac-seq-overlap-peaks-step',
	'/analysis-steps/kundaje-lab-atac-seq-peak-call-single-rep-v1/': 'kundaje-lab-atac-seq-peak-call-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-peak-call-v1/': 'kundaje-lab-atac-seq-peak-call-step',
	'/analysis-steps/kundaje-lab-atac-seq-pseudoreplicated-idr-peaks-conversion-single-rep-v1/': 'kundaje-lab-atac-seq-pseudoreplicated-idr-peaks-conversion-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-qc-single-rep-v1/': 'kundaje-lab-atac-seq-qc-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-qc-v1/': 'kundaje-lab-atac-seq-qc-step',
	'/analysis-steps/kundaje-lab-atac-seq-replicated-peaks-conversion-step-v1/': 'kundaje-lab-atac-seq-replicated-peaks-conversion-step',
	'/analysis-steps/kundaje-lab-atac-seq-signals-single-rep-v1/': 'kundaje-lab-atac-seq-signals-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-signals-v1/': 'kundaje-lab-atac-seq-signals-step',
	'/analysis-steps/kundaje-lab-atac-seq-stable-peaks-conversion-single-rep-v1/': 'kundaje-lab-atac-seq-stable-peaks-conversion-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-trim-align-filter-single-rep-v1/': 'kundaje-lab-atac-seq-trim-align-filter-single-rep-step',
	'/analysis-steps/kundaje-lab-atac-seq-trim-align-filter-v1/': 'kundaje-lab-atac-seq-trim-align-filter-step',
	'/analysis-steps/kundaje-lab-atac-seq-unreplicated-idr-single-rep-v1/': 'kundaje-lab-atac-seq-unreplicated-idr-single-rep-step',
	'/analysis-steps/index-rsem-v-1-0/': 'lrna-index-rsem-step',
	'/analysis-steps/index-star-v-1-0/': 'lrna-index-star-step',
	'/analysis-steps/index-tophat-v-1-0/': 'lrna-index-tophat-step',
	'/analysis-steps/lrna-pe-rsem-quantification-v-1/': 'lrna-pe-rsem-quantification-step',
	'/analysis-steps/lrna-pe-star-alignment-step-v-1/': 'lrna-pe-star-alignment-step',
	'/analysis-steps/lrna-pe-star-signal-step-v-1/': 'lrna-pe-star-signal-step',
	'/analysis-steps/lrna-pe-star-signals-for-tophat-step-v-1/': 'lrna-pe-star-signals-for-tophat-step',
	'/analysis-steps/lrna-pe-tophat-alignment-step-v-1/': 'lrna-pe-tophat-alignment-step',
	'/analysis-steps/long-rna-mad-qc-v-1/': 'lrna-repcorr-qc-pe-step',
	'/analysis-steps/long-rna-mad-qc-alt-v-1/': 'lrna-repcorr-qc-se-step',
	'/analysis-steps/lrna-se-rsem-quantification-step-v-1/': 'lrna-se-rsem-quantification-step',
	'/analysis-steps/lrna-se-star-alignment-step-v-1/': 'lrna-se-star-alignment-step',
	'/analysis-steps/lrna-se-star-signal-step-v-1/': 'lrna-se-star-signal-step',
	'/analysis-steps/lrna-se-star-signals-for-tophat-step-v-1/': 'lrna-se-star-signals-for-tophat-step',
	'/analysis-steps/lrna-se-tophat-alignment-step-v-1/': 'lrna-se-tophat-alignment-step',
	'/analysis-steps/step2-align-bowtie/': 'mango-align-bowtie-step',
	'/analysis-steps/step5-calc-interaction-confidence/': 'mango-calc-interaction-confidence-step',
	'/analysis-steps/step0-index-bowtie/': 'mango-index-bowtie-step',
	'/analysis-steps/step4-macs2-call-peaks/': 'mango-macs2-call-peaks-step',
	'/analysis-steps/step3-make-signal/': 'mango-make-signal-step',
	'/analysis-steps/step1-trim-linkers/': 'mango-trim-linkers-step',
	'/analysis-steps/micro-rna-alignment-step-v-1/': 'micro-rna-alignment-step',
	'/analysis-steps/microrna-quantification-step-v1/': 'microrna-quantification-step',
	'/analysis-steps/microrna-seq-alignment-v1/': 'microrna-seq-alignment-step',
	'/analysis-steps/microrna-signal-step-v1/': 'microrna-signal-step',
	'/analysis-steps/modern-bwa-alignment-step-v-1/': 'modern-bwa-alignment-step',
	'/analysis-steps/modern-bwa-indexing-step-v-1/': 'modern-bwa-indexing-step',
	'/analysis-steps/modern-chip-seq-control-normalized-signal-generation-step-v-1/': 'modern-chip-seq-control-normalized-signal-generation-step',
	'/analysis-steps/modern-chip-seq-filter-for-optimal-idr-peaks-step-v-1/': 'modern-chip-seq-filter-for-optimal-idr-peaks-step',
	'/analysis-steps/modern-chip-seq-optimal-idr-step-v-1/': 'modern-chip-seq-optimal-idr-step',
	'/analysis-steps/modern-chip-seq-optimal-idr-thresholded-peaks-to-bigbed-step-v-1/': 'modern-chip-seq-optimal-idr-thresholded-peaks-to-bigbed-step',
	'/analysis-steps/modern-chip-seq-peaks-to-bigbed-step-v-1/': 'modern-chip-seq-peaks-to-bigbed-step',
	'/analysis-steps/modern-chip-seq-read-depth-normalized-signal-generation-step-v-1/': 'modern-chip-seq-read-depth-normalized-signal-generation-step',
	'/analysis-steps/modern-chip-seq-replicate-alignment-pooling-step-v-1/': 'modern-chip-seq-replicate-alignment-pooling-step',
	'/analysis-steps/modern-chip-seq-replicate-pooled-control-normalized-signal-generation-step-v-1/': 'modern-chip-seq-replicate-pooled-control-normalized-signal-generation-step',
	'/analysis-steps/modern-chip-seq-replicate-pooled-read-depth-normalized-signal-generation-step-v-1/': 'modern-chip-seq-replicate-pooled-read-depth-normalized-signal-generation-step',
	'/analysis-steps/modern-chip-seq-replicate-pooled-unique-read-signal-generation-step-v-1/': 'modern-chip-seq-replicate-pooled-unique-read-signal-generation-step',
	'/analysis-steps/modern-chip-seq-spp-peak-calling-step-v-1/': 'modern-chip-seq-spp-peak-calling-step',
	'/analysis-steps/modern-chip-seq-unique-read-signal-generation-step-v-1/': 'modern-chip-seq-unique-read-signal-generation-step',
	'/analysis-steps/modern-transcript-alignment/': 'modern-transcript-alignment-step',
	'/analysis-steps/modern-transcript-quantification/': 'modern-transcript-quantification-step',
	'/analysis-steps/nanostring-bigbed-step-v-1/': 'nanostring-bigbed-step',
	'/analysis-steps/nanostring-mapping-step-v-1/': 'nanostring-mapping-step',
	'/analysis-steps/nanostring-nsolver-step-v-1/': 'nanostring-nsolver-step',
	'/analysis-steps/rampage-grit-peak-calling-step-v-1/': 'rampage-grit-peak-calling-step',
	'/analysis-steps/rampage-idr-step-v-1/': 'rampage-idr-step',
	'/analysis-steps/rampage-pe-alignment-step-v-1/': 'rampage-pe-alignment-step',
	'/analysis-steps/rampage-mad-qc-v-1/': 'rampage-repcorr-qc-step',
	'/analysis-steps/rampage-star-signal-step-v-1/': 'rampage-star-signal-step',
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-align-call-phase-step-v-1/': 'schatz-lab-variant-analysis-10x-longranger-align-call-phase-step',
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-index-step-v-1/': 'schatz-lab-variant-analysis-10x-longranger-index-step',
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-phased-diploid-genome-v1/': 'schatz-lab-variant-analysis-10x-longranger-phased-diploid-genome-step',
	'/analysis-steps/schatz-lab-variant-analysis-10x-vcf2diploid-phased-diploid-genome-v1/': 'schatz-lab-variant-analysis-10x-vcf2diploid-phased-diploid-genome-step',
	'/analysis-steps/schatz-lab-variant-analysis-gatk-indel-filtering-step-v1/': 'schatz-lab-variant-analysis-gatk-indel-filtering-step',
	'/analysis-steps/schatz-lab-variant-analysis-gatk-snp-filtering-step-v1/': 'schatz-lab-variant-analysis-gatk-snp-filtering-step',
	'/analysis-steps/schatz-lab-variant-analysis-gatk-variant-calling-v1/': 'schatz-lab-variant-analysis-gatk-variant-calling-step',
	'/analysis-steps/schatz-lab-variant-analysis-scalpel-filtering-step-v1/': 'schatz-lab-variant-analysis-scalpel-filtering-step',
	'/analysis-steps/schatz-lab-variant-analysis-scalpel-variant-calling-v1/': 'schatz-lab-variant-analysis-scalpel-variant-calling-step',
	'/analysis-steps/schatz-lab-variant-analysis-trim-bwa-mem-align-filter-v1/': 'schatz-lab-variant-analysis-trim-bwa-mem-align-filter-step',
	'/analysis-steps/schatz-lab-variant-analysis-trim-ngm-align-filter-v1/': 'schatz-lab-variant-analysis-trim-ngm-align-filter-step',
	'/analysis-steps/shrna-rna-seq-map-v1/': 'shrna-rna-seq-map-step',
	'/analysis-steps/shrna-rna-seq-signal-v1/': 'shrna-rna-seq-signal-step',
	'/analysis-steps/shrna-rna-splice-quant-v1/': 'shrna-rna-splice-quant-step',
	'/analysis-steps/small-rna-mad-qc-v-1/': 'small-rna-repcorr-qc-step',
	'/analysis-steps/small-rna-se-star-alignment-step-v-2/': 'small-rna-se-star-alignment-step',
	'/analysis-steps/small-rna-star-indexing-step-v-1/': 'small-rna-star-indexing-step',
	'/analysis-steps/small-rna-star-stranded-signal-step-v-2/': 'small-rna-star-stranded-signal-step',
	'/analysis-steps/tf-idr-peaks-to-bigbed-step/': 'tf-idr-peaks-to-bigbed-step',
	'/analysis-steps/tf-idr-step/': 'tf-idr-step',
	'/analysis-steps/tf-macs2-signal-calling-step/': 'tf-macs2-signal-calling-step',
	'/analysis-steps/tf-peaks-to-bigbed-step/': 'tf-peaks-to-bigbed-step',
	'/analysis-steps/tf-spp-peak-calling-step/': 'tf-spp-peak-calling-step',
	'/analysis-steps/tf-unreplicated-idr-peaks-to-bigbed-step/': 'tf-unreplicated-idr-peaks-to-bigbed-step',
	'/analysis-steps/tf-unreplicated-idr-step/': 'tf-unreplicated-idr-step',
	'/analysis-steps/tf-unreplicated-macs2-signal-calling-step/': 'tf-unreplicated-macs2-signal-calling-step',
	'/analysis-steps/tf-unreplicated-peaks-to-bigbed-step/': 'tf-unreplicated-peaks-to-bigbed-step',
	'/analysis-steps/tf-unreplicated-spp-peak-calling-step/': 'tf-unreplicated-spp-peak-calling-step',
	'/analysis-steps/wgbs-bismark-quantification-step-v-1/': 'wgbs-bismark-quantification-step',
	'/analysis-steps/wgbs-methylation-to-bigbed-conversion-step-v-1/': 'wgbs-methylation-to-bigbed-conversion-step'
}

major_version_mapping = {
	'/analysis-steps/alignment-pooling-step/': 1,
	'/analysis-steps/alignment-subsampling-step/': 1,
	'/analysis-steps/atac-seq-alignment-step-v-1/': 1,
	'/analysis-steps/atac-seq-peak-calling-step-v_1/': 1,
	'/analysis-steps/atac-seq-peak-filtering-step-v_1/': 1,
	'/analysis-steps/atac-seq-region-calling-step-v-1/': 1,
	'/analysis-steps/atac-seq-region-filter-step-v-1/': 1,
	'/analysis-steps/atac-seq-signal-gen-step-v-1/': 1,
	'/analysis-steps/bed-2-bigbed/': 1,
	'/analysis-steps/broad_frip_seq_cuffdiff_step/': 1,
	'/analysis-steps/broad_frip_seq_fastq_alignment_step/': 1,
	'/analysis-steps/bwa-alignment-step-v-1/': 1,
	'/analysis-steps/bwa-indexing-step-v-1/': 1,
	'/analysis-steps/bwa-raw-alignment-step-v-1/': 1,
	'/analysis-steps/celltype-cre-identification/': 1,
	'/analysis-steps/chia_pet_chromatin_interactions_calling_step/': 1,
	'/analysis-steps/chia_pet_haplotype_specific_step/': 1,
	'/analysis-steps/chia_pet_loop_calling_step/': 1,
	'/analysis-steps/chia_pet_peak_calling_step/': 1,
	'/analysis-steps/chia_pet_preprocessing_step/': 1,
	'/analysis-steps/cluster-filter/': 1,
	'/analysis-steps/control-alignment-subsampling-step/': 1,
	'/analysis-steps/dac-enhancer-like-ranking-method/': 1,
	'/analysis-steps/dac-promotor-like-ranking-method/': 1,
	'/analysis-steps/dme-align-se-v-2/': 2,
	'/analysis-steps/dme-index-bismark-v-2_deleted/': 2,
	'/analysis-steps/index-star-v-2-0/': 2,
	'/analysis-steps/lrna-pe-star-alignment-step-v-2-0/': 2,
	'/analysis-steps/lrna-pe-star-stranded-signal-step-v-2-0/': 2,
	'/analysis-steps/lrna-pe-star-stranded-signals-for-tophat-step-v-2-0/': 2,
	'/analysis-steps/lrna-se-star-alignment-step-v-2-0/': 2,
	'/analysis-steps/lrna-se-star-unstranded-signal-step-v-2-0/': 2,
	'/analysis-steps/lrna-se-star-unstranded-signals-for-tophat-step-v-2-0/': 2,
	'/analysis-steps/rampage-grit-peak-calling-step-v-1-1/': 1,
	'/analysis-steps/dme-align-pe-v-1/': 1,
	'/analysis-steps/mott-trim-align-bismark-v-1-0/': 1,
	'/analysis-steps/dme-extract-pe-v-1/': 1,
	'/analysis-steps/dme-extract-se-v-2/': 1,
	'/analysis-steps/wgbs-bismark-indexing-step-v-1/': 1,
	'/analysis-steps/dme-index-bismark-bowtie2-v-1/': 1,
	'/analysis-steps/dme-rep-corr-v-1/': 1,
	'/analysis-steps/dme-rep-corr-alt-v-1/': 1,
	'/analysis-steps/dnase-align-bwa-pe-v-1/': 1,
	'/analysis-steps/dnase-align-bwa-se-v-1/': 1,
	'/analysis-steps/dnase-bed-to-bigbed/': 1,
	'/analysis-steps/dnase-call-hotspots-v-1/': 1,
	'/analysis-steps/dnase-call-hotspots-alt-v-1/': 1,
	'/analysis-steps/dnase-eval-bam-v-1/': 1,
	'/analysis-steps/dnase-eval-bam-alt-v-1/': 1,
	'/analysis-steps/dnase-filter-pe-v-1/': 1,
	'/analysis-steps/dnase-filter-se-v-1/': 1,
	'/analysis-steps/dnase-index-bwa-v-1/': 1,
	'/analysis-steps/dnase-rep-corr-v-1/': 1,
	'/analysis-steps/dnase-rep-corr-alt-v-1/': 1,
	'/analysis-steps/dnase-seq-mapping-peak-calling-step-v-1/': 1,
	'/analysis-steps/dnase-seq-mapping-step-v2/': 2,
	'/analysis-steps/dnase-seq-peak-calling-step-v2/': 2,
	'/analysis-steps/eclip_bigbed_to_bed_step/': 1,
	'/analysis-steps/eclip_makebigwig_step/': 1,
	'/analysis-steps/eclip_normalization_step/': 1,
	'/analysis-steps/eclip_trimming_mapping_step/': 1,
	'/analysis-steps/encode2-dgf-analysis/': 1,
	'/analysis-steps/encode2-dnase-analysis/': 1,
	'/analysis-steps/generalized-cre-identification/': 1,
	'/analysis-steps/ggr_cl1_atac_seq_bam_to_bed_step/': 1,
	'/analysis-steps/ggr_cl1_atac_seq_idr_step/': 1,
	'/analysis-steps/ggr_cl1_atac_seq_mapping_step/': 1,
	'/analysis-steps/ggr_cl1_chip_seq_mapping_step/': 1,
	'/analysis-steps/ggr_cl1_chip_seq_quantification_step/': 1,
	'/analysis-steps/ggr_cl1_rna_seq_quantification_step/': 1,
	'/analysis-steps/ggr_cl1_rna_seq_trimming_mapping_filtering_step/': 1,
	'/analysis-steps/ggr_lg1_atac_seq_bet_to_bigwig_step/': 1,
	'/analysis-steps/ggr_lg1_atac_seq_fastq_align_n_filter_step/': 1,
	'/analysis-steps/ggr_lg1_atac_seq_peak_calling_step/': 1,
	'/analysis-steps/ggr_lg1_chip_seq_bam_to_bigwig_step/': 1,
	'/analysis-steps/ggr_lg1_chip_seq_fastq_alignment_step/': 1,
	'/analysis-steps/ggr_lg1_chip_seq_peak_calling_step/': 1,
	'/analysis-steps/ggr_lg1_rna_seq_bam2bigwig_step/': 1,
	'/analysis-steps/ggr_lg1_rna_seq_fastq_filtration_step/': 1,
	'/analysis-steps/ggr_lg1_rna_seq_mapping_step/': 1,
	'/analysis-steps/ggr_lg1_rna_seq_rsem_step/': 1,
	'/analysis-steps/ggr_lg2_rna_seq_bam2bigwig_step/': 1,
	'/analysis-steps/ggr_lg2_rna_seq_fastq_filtration_bowtie_step/': 1,
	'/analysis-steps/ggr_lg2_rna_seq_fastq_filtration_tophat_step/': 1,
	'/analysis-steps/ggr_lg2_rna_seq_rsem_step/': 1,
	'/analysis-steps/ggr_tr1_chip_seq_mapping_filtering_step/': 1,
	'/analysis-steps/ggr_tr1_chip_seq_peak_calling_step/': 1,
	'/analysis-steps/ggr_tr1_chip_seq_quantification_step/': 1,
	'/analysis-steps/ggr_tr1_dnase_seq_mapping_filtering_step/': 1,
	'/analysis-steps/ggr_tr1_dnase_seq_peak_calling_step/': 1,
	'/analysis-steps/ggr_tr1_dnase_seq_quantification_step/': 1,
	'/analysis-steps/ggr_tr1_hic_loop_calling_step/': 1,
	'/analysis-steps/ggr_tr1_hic_mapping_step/': 1,
	'/analysis-steps/ggr_tr1_hic_tad_calling_step/': 1,
	'/analysis-steps/ggr_tr1_rna_seq_gene_quantification_step/': 1,
	'/analysis-steps/ggr_tr1_rna_seq_signal_generation_step/': 1,
	'/analysis-steps/ggr_tr1_rna_seq_trimming_mapping_step/': 1,
	'/analysis-steps/hic-liftover-step/': 1,
	'/analysis-steps/histone-overlap-peaks-step-v-1/': 1,
	'/analysis-steps/histone-peak-calling-step-v-1/': 1,
	'/analysis-steps/histone-peaks-to-bigbed-step-v-1/': 1,
	'/analysis-steps/histone-replicated-peaks-to-bigbed-step-v-1/': 1,
	'/analysis-steps/histone-unreplicated-partition-concordance-peaks-to-bigbed-step/': 1,
	'/analysis-steps/histone-unreplicated-partition-concordance-step/': 1,
	'/analysis-steps/histone-unreplicated-peak-calling-step/': 1,
	'/analysis-steps/histone-unreplicated-peaks-to-bigbed-step/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-filtered-peaks-conversion-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-filtered-peaks-conversion-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-idr-peaks-conversion-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-idr-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-overlap-peaks-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-overlap-peaks-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-peak-call-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-peak-call-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-pseudoreplicated-idr-peaks-conversion-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-qc-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-qc-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-replicated-peaks-conversion-step-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-signals-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-signals-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-stable-peaks-conversion-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-trim-align-filter-single-rep-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-trim-align-filter-v1/': 1,
	'/analysis-steps/kundaje-lab-atac-seq-unreplicated-idr-single-rep-v1/': 1,
	'/analysis-steps/index-rsem-v-1-0/': 1,
	'/analysis-steps/index-star-v-1-0/': 1,
	'/analysis-steps/index-tophat-v-1-0/': 1,
	'/analysis-steps/lrna-pe-rsem-quantification-v-1/': 1,
	'/analysis-steps/lrna-pe-star-alignment-step-v-1/': 1,
	'/analysis-steps/lrna-pe-star-signal-step-v-1/': 1,
	'/analysis-steps/lrna-pe-star-signals-for-tophat-step-v-1/': 1,
	'/analysis-steps/lrna-pe-tophat-alignment-step-v-1/': 1,
	'/analysis-steps/long-rna-mad-qc-v-1/': 1,
	'/analysis-steps/long-rna-mad-qc-alt-v-1/': 1,
	'/analysis-steps/lrna-se-rsem-quantification-step-v-1/': 1,
	'/analysis-steps/lrna-se-star-alignment-step-v-1/': 1,
	'/analysis-steps/lrna-se-star-signal-step-v-1/': 1,
	'/analysis-steps/lrna-se-star-signals-for-tophat-step-v-1/': 1,
	'/analysis-steps/lrna-se-tophat-alignment-step-v-1/': 1,
	'/analysis-steps/step2-align-bowtie/': 1,
	'/analysis-steps/step5-calc-interaction-confidence/': 1,
	'/analysis-steps/step0-index-bowtie/': 1,
	'/analysis-steps/step4-macs2-call-peaks/': 1,
	'/analysis-steps/step3-make-signal/': 1,
	'/analysis-steps/step1-trim-linkers/': 1,
	'/analysis-steps/micro-rna-alignment-step-v-1/': 1,
	'/analysis-steps/microrna-quantification-step-v1/': 1,
	'/analysis-steps/microrna-seq-alignment-v1/': 1,
	'/analysis-steps/microrna-signal-step-v1/': 1,
	'/analysis-steps/modern-bwa-alignment-step-v-1/': 1,
	'/analysis-steps/modern-bwa-indexing-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-control-normalized-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-filter-for-optimal-idr-peaks-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-optimal-idr-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-optimal-idr-thresholded-peaks-to-bigbed-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-peaks-to-bigbed-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-read-depth-normalized-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-replicate-alignment-pooling-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-replicate-pooled-control-normalized-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-replicate-pooled-read-depth-normalized-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-replicate-pooled-unique-read-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-spp-peak-calling-step-v-1/': 1,
	'/analysis-steps/modern-chip-seq-unique-read-signal-generation-step-v-1/': 1,
	'/analysis-steps/modern-transcript-alignment/': 1,
	'/analysis-steps/modern-transcript-quantification/': 1,
	'/analysis-steps/nanostring-bigbed-step-v-1/': 1,
	'/analysis-steps/nanostring-mapping-step-v-1/': 1,
	'/analysis-steps/nanostring-nsolver-step-v-1/': 1,
	'/analysis-steps/rampage-grit-peak-calling-step-v-1/': 1,
	'/analysis-steps/rampage-idr-step-v-1/': 1,
	'/analysis-steps/rampage-pe-alignment-step-v-1/': 1,
	'/analysis-steps/rampage-mad-qc-v-1/': 1,
	'/analysis-steps/rampage-star-signal-step-v-1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-align-call-phase-step-v-1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-index-step-v-1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-10x-longranger-phased-diploid-genome-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-10x-vcf2diploid-phased-diploid-genome-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-gatk-indel-filtering-step-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-gatk-snp-filtering-step-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-gatk-variant-calling-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-scalpel-filtering-step-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-scalpel-variant-calling-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-trim-bwa-mem-align-filter-v1/': 1,
	'/analysis-steps/schatz-lab-variant-analysis-trim-ngm-align-filter-v1/': 1,
	'/analysis-steps/shrna-rna-seq-map-v1/': 1,
	'/analysis-steps/shrna-rna-seq-signal-v1/': 1,
	'/analysis-steps/shrna-rna-splice-quant-v1/': 1,
	'/analysis-steps/small-rna-mad-qc-v-1/': 1,
	'/analysis-steps/small-rna-se-star-alignment-step-v-2/': 1,
	'/analysis-steps/small-rna-star-indexing-step-v-1/': 1,
	'/analysis-steps/small-rna-star-stranded-signal-step-v-2/': 1,
	'/analysis-steps/tf-idr-peaks-to-bigbed-step/': 1,
	'/analysis-steps/tf-idr-step/': 1,
	'/analysis-steps/tf-macs2-signal-calling-step/': 1,
	'/analysis-steps/tf-peaks-to-bigbed-step/': 1,
	'/analysis-steps/tf-spp-peak-calling-step/': 1,
	'/analysis-steps/tf-unreplicated-idr-peaks-to-bigbed-step/': 1,
	'/analysis-steps/tf-unreplicated-idr-step/': 1,
	'/analysis-steps/tf-unreplicated-macs2-signal-calling-step/': 1,
	'/analysis-steps/tf-unreplicated-peaks-to-bigbed-step/': 1,
	'/analysis-steps/tf-unreplicated-spp-peak-calling-step/': 1,
	'/analysis-steps/wgbs-bismark-quantification-step-v-1/': 1,
	'/analysis-steps/wgbs-methylation-to-bigbed-conversion-step-v-1/': 1
}

title_mapping = {
    '/analysis-steps/dme-index-bismark-v-2_deleted/': 'WGBS Bismark-bowtie1 genome indexing step - Version 2',
    '/analysis-steps/lrna-pe-star-alignment-step-v-2-0/': 'Long RNA-seq STAR paired-ended alignment step v2.0',
    '/analysis-steps/lrna-se-star-alignment-step-v-2-0/': 'Long RNA-seq STAR single-ended alignment step v2.0',
    '/analysis-steps/dme-index-bismark-bowtie2-v-1/': 'WGBS Bismark-bowtie2 genome indexing step - Version 1',
    '/analysis-steps/dnase-seq-mapping-peak-calling-step-v-1/': 'DNase-seq mapping and peak calling step'
}

status_mapping = {
	'/analysis-steps/alignment-pooling-step/': 'released',
	'/analysis-steps/alignment-subsampling-step/': 'released',
	'/analysis-steps/dme-extract-se-v-2/': 'released',
	'/analysis-steps/dme-rep-corr-alt-v-1/': 'released',
	'/analysis-steps/dme-rep-corr-v-1/': 'released',
	'/analysis-steps/modern-chip-seq-filter-for-optimal-idr-peaks-step-v-1/': 'released',
	'/analysis-steps/modern-chip-seq-optimal-idr-thresholded-peaks-to-bigbed-step-v-1/': 'released'
}

aliases_mapping = {
	'/analysis-steps/alignment-pooling-step/': 'encode:alignment-pooling-step-v-1',
	'/analysis-steps/alignment-subsampling-step/': 'encode:alignment-subsampling-step-v-1',
	'/analysis-steps/broad_frip_seq_fastq_alignment_step/': 'encode:frip-seq-pipeline-alignment-step-v-1',
	'/analysis-steps/control-alignment-subsampling-step/': 'encode:control-alignment-subsampling-step-v-1',
	'/analysis-steps/dac-enhancer-like-ranking-method/': 'zhiping-weng:dac-enhancer-like-ranking-method-step-v-1',
	'/analysis-steps/dac-promotor-like-ranking-method/': 'zhiping-weng:dac-promoter-like-ranking-method-step-v-1',
	'/analysis-steps/dme-index-bismark-v-2_deleted/': 'dnanexus:deleted-dme-index-bismark-step-v-2',
	'/analysis-steps/dnase-bed-to-bigbed/': 'john-stamatoyannopoulos:dnase-bed-to-bigbed-v-1',
	'/analysis-steps/dnase-seq-mapping-peak-calling-step-v-1/': 'john-stamatoyannopoulos:dnase-seq-mapping-peak-calling-step-v-1',
	'/analysis-steps/dnase-seq-mapping-step-v2/': 'john-stamatoyannopoulos:dnase-seq-mapping-step-v-2',
	'/analysis-steps/dnase-seq-peak-calling-step-v2/': 'john-stamatoyannopoulos:dnase-seq-peak-calling-step-v-2',
	'/analysis-steps/eclip_makebigwig_step/': 'gene-yeo:eclip-makebigwig-step-v-1',
	'/analysis-steps/encode2-dgf-analysis/': 'john-stamatoyannopoulos:encode2-dgf-analysis',
	'/analysis-steps/encode2-dnase-analysis/': 'john-stamatoyannopoulos:encode2-dnase-analysis',
	'/analysis-steps/hic-liftover-step/': 'zhiping-weng:hic-liftover-step-v-1',
	'/analysis-steps/index-star-v-2-0/': 'encoded:deleted-lrna-index-star-v-2',
	'/analysis-steps/lrna-pe-star-alignment-step-v-2-0/': 'encode:deleted-lrna-pe-star-alignment-step-v-2',
	'/analysis-steps/lrna-pe-star-stranded-signal-step-v-2-0/': 'encode:deleted-lrna-pe-star-stranded-signal-step-v-2',
	'/analysis-steps/lrna-pe-star-stranded-signals-for-tophat-step-v-2-0/': 'encode:deleted-lrna-pe-star-stranded-signals-for-tophat-step-v-2',
	'/analysis-steps/lrna-se-star-alignment-step-v-2-0/': 'encode:deleted-lrna-se-star-alignment-step-v-2',
	'/analysis-steps/lrna-se-star-unstranded-signal-step-v-2-0/': 'encode:deleted-lrna-se-star-unstranded-signal-step-v-2',
	'/analysis-steps/lrna-se-star-unstranded-signals-for-tophat-step-v-2-0/': 'encode:deleted-lrna-se-star-unstranded-signals-for-tophat-step-v-2',
	'/analysis-steps/micro-rna-alignment-step-v-1/': 'ali-mortazavi:micro-rna-alignment-step-v-1',
	'/analysis-steps/nanostring-mapping-step-v-1/': 'ali-mortazavi:nanostring-mapping-step-v-1',
	'/analysis-steps/rampage-grit-peak-calling-step-v-1-1/': 'encode:deleted-rampage-grit-peak-calling-step-v-1',
	'/analysis-steps/shrna-rna-seq-map-v1/': 'encode:shrna-rna-seq-map-step-v-1',
	'/analysis-steps/shrna-rna-seq-signal-v1/': 'encode:shrna-rna-seq-signal-step-v-1',
	'/analysis-steps/shrna-rna-splice-quant-v1/': 'encode:shrna-rna-splice-quant-step-v-1'
}
